### ISSUE TITLE
Fava freezes on request inside Docker container (only!)

### ISSUE THREAD

**Issue by umangv** [beancount/fava#1792]
I have only been able to reproduce this when running Fava when all the following are true:
 1. Fava is running in a Docker container 
 2. Fava is running with a production HTTP server (I tried the default cheroot and also tried running with gunicorn), not reproducible running Fava with the Flask dev server

I've [narrowed it down to the watcher thread](https://github.com/beancount/fava/blob/7fc834f2020a86244cb5709b0757ebdc0cd2613f/src/fava/core/__init__.py#L284):
```
        if not self._is_encrypted:
            self.watcher.update(*self.paths_to_watch())
```

Simply deleting that line makes the hanging go away.

I can trace the hanging to [pre-request initialization (`fava.application._setup_filters._pull_beancount_file`, where `FavaLedger` is created)](https://github.com/beancount/fava/blob/7fc834f2020a86244cb5709b0757ebdc0cd2613f/src/fava/application.py#L231).

My initial suspicion is that this is some kind of subtle deadlock/race condition because it only appears in certain circumstances. My Python-specific multi-threading experience is limited, so I don't know all the tools available, but I'm going to keep poking.

To reproduce in a Docker container, here's a Dockerfile:

```
# syntax=docker/dockerfile:1
FROM python:3.10-slim
ENV PYTHONUNBUFFERED 1

RUN apt update && apt install -y git npm

RUN python -m venv /opt/venv
ENV VIRTUAL_ENV="/opt/venv"
ENV PATH="/opt/venv/bin:$PATH"

WORKDIR /bean
RUN mkdir -p beancount smart_importer fava

RUN git clone --branch v2 https://github.com/beancount/beancount.git
RUN git clone https://github.com/beancount/fava.git

RUN pip install ./beancount
RUN pip install ./fava[dev]

# Uncomment these to override what gets packaged into the image (to experiment with local changes)
# Make sure that these are these same as HEAD + intended local changes).
# COPY fava/src/fava/core/watcher.py /opt/venv/lib/python3.10/site-packages/fava/core/watcher.py
# COPY fava/src/fava/core/__init__.py /opt/venv/lib/python3.10/site-packages/fava/core/__init__.py

RUN bean-example > /example.bean
CMD ["fava", "/example.bean", "--port", "5000", "--host", "0.0.0.0"]
```

**Comment by umangv** [comment]
Searching online, I found `pystack` and I ran `pystack remote ${FAVA_PID} --locals --no-block` before any requests came in (to see the idle state) and after. Here are my main findings:

First, there is a new thread running the watcher (as expected). 

```
Traceback for thread 118 (fava) [Has the GIL] (most recent call last):
    (Python) File "/usr/local/lib/python3.10/threading.py", line 973, in _bootstrap
        self._bootstrap_inner()
      Arguments:
        self: <_WatchfilesThread at 0x733bae5000d0>
    (Python) File "/usr/local/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
        self.run()
      Arguments:
        self: <_WatchfilesThread at 0x733bae5000d0>
    (Python) File "/opt/venv/lib/python3.10/site-packages/fava/core/watcher.py", line 60, in run
        for changes in watch(
      Arguments:
        self: <_WatchfilesThread at 0x733bae5000d0>
      Locals:
        watch_paths: <set at 0x733bcc04a260>
    (Python) File "/opt/venv/lib/python3.10/site-packages/watchfiles/main.py", line 118, in watch
        with RustNotify(
      Locals:
        recursive: True
        poll_delay_ms: 300
        raise_interrupt: True
        debug: False
        paths: (<PosixPath at 0x733bcc0ed3c0>)
        ignore_permission_denied: True
        yield_on_timeout: False
        rust_timeout: 5000
        force_polling: False
        step: 50
        stop_event: <Event at 0x733bae551750>
        debounce: 1600
        watch_filter: <method at 0x733bcfb17000>
```

What is surprising is that it has the GIL. From [the watchfiles documentation](https://watchfiles.helpmanual.io/#how-watchfiles-works):

> watchfiles is based on the [Notify](https://github.com/notify-rs/notify) rust library.
> 
> All the hard work of integrating with the OS's file system events notifications and falling back to polling is palmed off onto the rust library.

It seems like in some conditions (running in a Docker container with cheroot), watchfiles either hangs onto the GIL when there are no changes or the actual call to `RustNotify` hangs (which causes the GIL never to be released). I'm suspecting it's the latter, because I tried adding a `print("Saw change")` to the `for changes in watch(...):` loop in in `fava.core.watcher` and then (after fava seemed to have hung) changing an innocuous line in `/example.bean`, but fava seemed to make no progress. So even if `watchfiles` is just hanging on to the GIL, you'd expect that "Saw change" would be printed. Instead, given that nothing happens, I suspect that `RustNotify` is hanging.

The only other changes to the `pystack` output were the thread where fava code was being executed (it seemed to be stuck on one of the `print()` lines I added to debug) and the main thread (which was running cheroot code and it now had a couple of new local variables: `new_conn` and a `sock_fd`).

All of this is making me wonder what is happening in `watchfiles`, especially inside `RustNotify`. Will investigate more when I get a chance. I'll also look into seeing whether it's possible to switch to pure-Python filesystem polling, which might be more inefficient but at least we'll know that the thread will release the GIL.

**Comment by umangv** [comment]
Gah. This was much less interesting than I thought!

I was able to fix this by passive `False` for [`recursive` in `watchfiles.watch`](https://github.com/samuelcolvin/watchfiles/blob/9f19a18da5323d0e63a2ee16c47ed6c819f89be5/watchfiles/main.py#L101-L102) in [fava code](https://github.com/beancount/fava/blob/7fc834f2020a86244cb5709b0757ebdc0cd2613f/src/fava/core/watcher.py#L51-L56). Which made me want to check if `recursive` is somehow operating on the parent directory rather than the file itself. I can't find the code that does this, but it definitely looks like this is the behavior. Moving the file to a different directory (e.g., `mv /example.bean /bean/data/example.bean`) fixed the problem.

It's not yet clear to me if it actually makes sense for fava to switch to `recursive=False` for everyone (since I assume the only paths being passed in are files and not directories), but at least I have a quick fix.

**Comment by umangv** [comment]
Oh, the confusing behavior happens because of [the following line](https://github.com/beancount/fava/blob/7fc834f2020a86244cb5709b0757ebdc0cd2613f/src/fava/core/watcher.py#L49C9-L49C76):

```
watch_paths = {f.parent for f in self._file_set} | self._folder_set
```

If this is the intended behavior, then it kind of makes sense that the server would come to halt as soon as we try to watch `/`. Not sure how to prevent others from facing the same problem as I have, but I also don't have any suggestion on what fava could be doing differently.

**Comment by umangv** [comment]
I guess the only reasonable thing I can think of is to print a warning when watching a directory that transitively contains more than a fixed number (e.g., 1,000) files, which seems hard to implement, or when the directory is one of a few known directories (e.g., `/` on *nix and `C:/` on Windows), but that seems like a terrible band-aid for this problem.

I'm going to close this issue. @yagebu (who is, to my knowledge, the primary maintainer of fava), feel free to re-open this if you think it's worthy of doing something about.

**Comment by yagebu** [comment]
Thanks for reporting and debugging this.

I think this is something we should fix - the recursive watching of the parent directory of the beancount file is an unintended side effect of the fix for #1778. `watchfiles` only allows specifying whether to recursively watch once - and for the document directories that's something we want. Simplest way to fix this is to span watchfiles separately for files and folders.

**Comment by umangv** [comment]
Thanks, @yagebu! I rebuilt the Docker image with #1794 and I confirmed that it doesn't hang.
